### PR TITLE
Add aioapns version of APNS

### PR DESCRIPTION
**Pull request by pomali**
Since version python 3.10+ is not supported by `apns2` (because of dependency on `hyper`) we add `aioapns` version of sending APNs notifications.

We add installation extra `[APNS_ASYNC]` that installs aioapns and use new version of service if `aioapns` is installed. Tests are also conditional on installed modules/version of python.

Related to 
- https://github.com/jazzband/django-push-notifications/pull/623 (is blocked by apns2)
- https://github.com/jazzband/django-push-notifications/issues/685
- https://github.com/jazzband/django-push-notifications/issues/622
- https://github.com/jazzband/django-push-notifications/issues/675

### CODE REVIEW

**Comment by pomali** [comment]
There are some (maybe) controversial decisions, feel free to discuss them
- choice of `aioapns` - it looks like only library with active development
- supporting both `apns2` and `aioapns` - because we don't want to break older code
- deciding on which version to use in `models.py` - sure we can move it to `apns.py` and move apns2 code to `apns_legacy.py`

**Comment by sivasankarankb** [comment]
@pomali You have left `print()` statements in the code. See `apns_async.py`.

**Comment by itayAmza** [comment]
@pomali, any progress for this one?
I had the [apns2 0.7.2 requires PyJWT<2.0.0,>=1.4.0, but you have pyjwt 2.1.0 which is incompatible](https://github.com/Pr0Ger/PyAPNs2/issues/133) issue, and pointing to your branch (as a temporary solution) completely solves it.

**Comment by pomali** [comment]
I'm not sure what's missing, maybe someone should review it? And then there is code coverage check failing, but most of failing lines are comments, so I am not sure how important it is. 

Is there anyone who could point me in right direction?

@itayAmza  glad that it helped :)

**Review by VaughnDV — APPROVED** [review]

**Comment by 50-Course** [comment]
hopefully, I get to have a look into both PRs, tonight. Thank you for the notification @itayAmza

**Comment by jamaalscarlett** [comment]
1. I can review the PR, but I have never used APNS.  Has someone actually run this with an ios/macos app?
2. Regarding the support for both apns2 and aioapns, I think we should add a deprecation warning with this release then completely remove apns2 with the next release.

**Comment by pomali** [comment]
I ran it with an iOS simulator, but a second pair of eyes with a real device would be good idea

**Comment by ssyberg** [comment]
> I ran it with an iOS simulator, but a second pair of eyes with a real device would be good idea

I just tried this and got this error when trying to send a test message via apns:

```
ModuleNotFoundError at /admin/push_notifications/apnsdevice/
No module named 'apns2'
Request Method:	POST
Request URL:	https://local.a24-vod:8060/admin/push_notifications/apnsdevice/
Django Version:	4.2
Exception Type:	ModuleNotFoundError
Exception Value:	
No module named 'apns2'
```

File: push_notifications/apns_async.py
Comment (kupsum) [inline push_notifications/apns_async.py:149]: Since `notification_request_kwargs` argument has a default value and it's content is changed in the function logic, the default state can change during execution. If any previous call has set values `expiration`, `priority` or `collapse_id`, these are persisted to any next call and if the next call  does not have these values or has the value set to `None`, the values are still passed to sent `NotificationRequest`. A simple test case to demonstrate this is to send first message with `expiration` and second without `expiration` set. The value from first call is passed to second call `NotificationRequest`.

**Comment by kupsum** [comment]
There is a possible bug with default argument pollution. Below test case can demonstrate the issue. I've also added a review comment about this.

```
	@mock.patch("push_notifications.apns_async.APNs", autospec=True)
	def test_apns_send_messages_different_priority(self, mock_apns):
		self._create_devices(["abc", "def"])
		device_1 = APNSDevice.objects.get(registration_id="abc")
		device_2 = APNSDevice.objects.get(registration_id="def")

		device_1.send_message("Hello world 1", expiration=time.time() + 1, priority=5, collapse_id="1")
		args_1, _ = mock_apns.return_value.send_notification.call_args

		device_2.send_message("Hello world 2")
		args_2, _ = mock_apns.return_value.send_notification.call_args

		req = args_1[0]
		self.assertEqual(req.device_token, "abc")
		self.assertEqual(req.message["aps"]["alert"], "Hello world 1")
		self.assertAlmostEqual(req.time_to_live, 1, places=-1)
		self.assertEqual(req.priority, 5)
		self.assertEqual(req.collapse_key, "1")

		reg_2 = args_2[0]
		self.assertEqual(reg_2.device_token, "def")
		self.assertEqual(reg_2.message["aps"]["alert"], "Hello world 2")
		self.assertIsNone(reg_2.time_to_live, "No time to live should be specified")
		self.assertIsNone(reg_2.priority, "No priority should be specified")
		self.assertIsNone(reg_2.collapse_key, "No collapse key should be specified")

```

**Comment by pomali** [comment]
> > I ran it with an iOS simulator, but a second pair of eyes with a real device would be good idea
> 
> I just tried this and got this error when trying to send a test message via apns:
> 
> ```
> ModuleNotFoundError at /admin/push_notifications/apnsdevice/
> No module named 'apns2'
> Request Method:	POST
> Request URL:	https://local.a24-vod:8060/admin/push_notifications/apnsdevice/
> Django Version:	4.2
> Exception Type:	ModuleNotFoundError
> Exception Value:	
> No module named 'apns2'
> ```

do you have `aioapns` installed? because that is condition on which we use the new code

added mention to README
```
pip install django-push-notifications[APNS_ASYNC]
```

**Comment by pomali** [comment]
> There is a possible bug with default argument pollution. Below test case can demonstrate the issue. I've also added a review comment about this.
> 

Thanks for catching this, I added your test and fixed the issue.

**Comment by kmasuhr** [comment]
I think it is worth to add python 3.10 as testing target in CI, since it should enable support for version 3.10
https://github.com/jazzband/django-push-notifications/blob/820262efedbd71d38e62320b561c4aaed95b1131/.github/workflows/test.yml#L12

**Comment by ssyberg** [comment]
> > > I ran it with an iOS simulator, but a second pair of eyes with a real device would be good idea
> > 
> > 
> > I just tried this and got this error when trying to send a test message via apns:
> > ```
> > ModuleNotFoundError at /admin/push_notifications/apnsdevice/
> > No module named 'apns2'
> > Request Method:	POST
> > Request URL:	https://local.a24-vod:8060/admin/push_notifications/apnsdevice/
> > Django Version:	4.2
> > Exception Type:	ModuleNotFoundError
> > Exception Value:	
> > No module named 'apns2'
> > ```
> 
> do you have `aioapns` installed? because that is condition on which we use the new code
> 
> added mention to README
> 
> ```
> pip install django-push-notifications[APNS_ASYNC]
> ```

I'm pretty sure I did when I tested it, but wouldn't that be handled by pip/piptools?

**Comment by 50-Course** [comment]
Hello @kmasuhr,

I have just updated this PR to bump the version of python being used in CI.
However, it would seem one of its transitive dependencies is not compatible with Python 3.10 - precisely, the `hyper==0.7.0`
HTTP Client.

Given that the `hyper` library is not actively maintained, I would like to propose that we replace it with a maintained alternative.
I am open to suggestions, but I would recommend using `httpx` as a good replacement.

**Comment by pomali** [comment]
> I'm pretty sure I did when I tested it, but wouldn't that be handled by pip/piptools?
@ssyberg it should be handled by pip, but it's optional dependency, so maybe it didn't get installed?

**Comment by rib3** [comment]
> I have just updated this PR to bump the version of python being used in CI. However, it would seem one of its transitive dependencies is not compatible with Python 3.10 - precisely, the `hyper==0.7.0` HTTP Client.
> 
> Given that the `hyper` library is not actively maintained, I would like to propose that we replace it with a maintained alternative. I am open to suggestions, but I would recommend using `httpx` as a good replacement.

hyper is a dependency of `apns2`.
This PR adds support  for `aioapns` as an alternative apple push notification library.
(because of the problems with apns2/hyper)

There are some github issues with additional details / background linked in the PR description.

**Comment by ssyberg** [comment]
> > I have just updated this PR to bump the version of python being used in CI. However, it would seem one of its transitive dependencies is not compatible with Python 3.10 - precisely, the `hyper==0.7.0` HTTP Client.
> > Given that the `hyper` library is not actively maintained, I would like to propose that we replace it with a maintained alternative. I am open to suggestions, but I would recommend using `httpx` as a good replacement.
> 
> hyper is a dependency of `apns2`. This PR adds support for `aioapns` as an alternative apple push notification library. (because of the problems with apns2/hyper)
> 
> There are some github issues with additional details / background linked in the PR description.

If I'm understanding correctly, the inclusion of `hyper` in this PR is only to stay backwards compatible with `apns2` for people who don't want to upgrade to using `aioapns` in which case they need to stay <3.10 (which if they are using `apns2` I think is implied). IMO this could be noted in the docs but shouldn't block merge, a lot of people are waiting for this to get merged and it's blocking dev for anyone on later version setups.

File: tox.ini
Comment (kupsum) [inline tox.ini:7]: Django 4.2 supports still Python 3.8
```suggestion
 py{38,39,310,311}-dj{40,405,42}
```

**Review by kupsum — COMMENTED** [review]
A few comments when trying to install this in Python 3.11 system

File: tox.ini
Comment (pomali) [reply tox.ini:49]: updated deps, removed apns2 from default leaving it only in py{36,37,38,39}

File: tox.ini
Comment (pomali) [reply tox.ini:7]: accepted change

File: .github/workflows/test.yml
Comment (pomali) [reply .github/workflows/test.yml:12]: accepted change

File: setup.cfg
Comment (pomali) [reply setup.cfg:25]: accepted change

File: push_notifications/apns_async.py
Comment (pomali) [reply push_notifications/apns_async.py:149]: Solved it by creating copy of `notification_request_kwargs`, but if you consider it code smell I could rework it to having default argument `None` and replacing them in function with `{}`. Added your test (`test_apns_send_messages_different_priority`)

File: tests/test_apns_models.py
Comment (pomali) [reply tests/test_apns_models.py:15]: accepted change

File: tests/test_apns_push_payload.py
Comment (pomali) [reply tests/test_apns_push_payload.py:11]: accepted change

File: push_notifications/apns_async.py
Comment (50-Course) [inline push_notifications/apns_async.py:12]: Out of office right now, `Awaitable` is that some form of `Future` result?

Comment (50-Course) [reply push_notifications/apns_async.py:12]: #Update: Interesting, 

https://docs.python.org/3/library/typing.html#annotating-callable-objects

I have always take a different approach to this!

File: push_notifications/apns_async.py
Comment (50-Course) [inline push_notifications/apns_async.py:174]: ruff botched this formatting in here

File: push_notifications/apns_async.py
Comment (50-Course) [inline push_notifications/apns_async.py:1]: 👍🏼 looks good to me, can we have a test for this?

**Comment by ficast** [comment]
Hey guys, any news on this? Can we get this merged?

**Comment by illymarev** [comment]
Hello, are there any updates regarding this issue? Any other changes are needed to merge the PR?

**Comment by pomali** [comment]
I don't see any outstanding issues from my side.

**Comment by 50-Course** [comment]
haven't followed up much on the development lately, however it would seem we having some conflicts, no?

**Review by 50-Course — COMMENTED** [review]
PR looks greenlit, still on the lookout what's causing the rebase conflicts.

**Comment by illymarev** [comment]
Have you determined what causes rebase conflicts? Do you need any additional changes?

**Comment by DataGreed** [comment]
Anyone maintaining this package?.. its not usable at the moment on python 3.10+

**Review by nossila — APPROVED** [review]

**Review by ficast — APPROVED** [review]

**Comment by ficast** [comment]
let's merge this guys !!!

**Comment by 50-Course** [comment]
> Have you determined what causes rebase conflicts? Do you need any additional changes?

yes i just did - its national holiday over here and @ficast notification actually reminded me of the PR. currently fixing them conflicts.

**Comment by vitorguima** [comment]
@50-Course will this PR get merged by any chances?

**Review by vitorguima — APPROVED** [review]

**Comment by 50-Course** [comment]
Yes, am currently fixing the re-base conflicts a couple more changes to go... hopefully i after my office hours i could get to it again.

@vitorguima

**Comment by 50-Course** [comment]
Hi everyone,

I hope this message finds you well. I wanted to provide an update regarding this PR that led to a erroneous closure of the PR.

Unfortunately, while attempting to assist with the updates and re-base conflicts, I inadvertently pushed changes that led to the closure of the original pull request. I sincerely apologize for this error and any confusion it may have caused. My intention was to facilitate the merging of the contributions into the project, and I regret any inconvenience this has created.

To rectify the situation, I have now pushed the latest changes back to the `pomali/master` branch with proper attribution to the original author. I will be creating a new pull request shortly to ensure that these contributions can be reviewed and merged appropriately.

Thank you for your understanding, and I appreciate your patience during this process. Please feel free to reach out if you have any questions or need further assistance.

regards,
E.A

**Comment by 50-Course** [comment]
Quick heads-up,

I want to provide an update regarding the changes. Please follow the new PR on #739. The new changes have been fully re-based onto this PR to facilitate the merging process.

Thank you for your patience, and I appreciate your understanding as we work to get this merged.

**Comment by 50-Course** [comment]
Once again calling on review of the new PR as its now able to get merged. This would allow us successfully close this one out

**Comment by 50-Course** [comment]
Update to all following this thread, PR been merged. Documentation could use some work here and there due to rebase issues

**Pull request closed without merging**
